### PR TITLE
Adds an image publisher and velocity subscriber to Unity RosSharp

### PR DESCRIPTION
**Pull request by jeremyfix**
I propose you an implementation of : 
1) an ImagePublisher which is publishing a JPEG compressed image with parameters such as topic, compression rate, parametrised from Unity 

2) a velocity subscriber which is listening for a Twist and its companion TransformManager which is moving the object given the linear/angular velocities in the message

For the moment the TransformManager is directly applying a translation/rotation on the object. It does not control the position with Force/Torque so this is probably not going to work for now if you have a RigidBody with collision I think.

### CODE REVIEW

File: Unity3D/Assets/RosSharp/Scripts/ImagePublisher.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+﻿using System.Collections;
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/ImagePublisher.cs]: Please split the class ImagePublisher into the following 2 classes:
1 CameraImageManager:
* requires component typeof Camera
* calls CameraImagePublisher.Publish(Texture2D texture2D) every Update()

2 CameraImageManager
* requires component typeof RosConnector
* provides Public method Publish(Texture2D texture2D)

File: Unity3D/Assets/RosSharp/Scripts/ImagePublisher.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+using System.Collections.Generic;
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/ImagePublisher.cs]: Please rename the variables as follows:
cameraToPublish -> camera
baseTopic -> topic
resWidth -> resolutionWidth
resHeight -> resolutionHeight
pub_id -> topicID or publicationID
rt -> renderTexture
viewport -> rect
screenshot -> texture2D

File: Unity3D/Assets/RosSharp/Scripts/ImagePublisher.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+using UnityEngine;
+
+namespace RosSharp.RosBridgeClient
+{
+	[RequireComponent(typeof(RosConnector))]
+	public class ImagePublisher : MonoBehaviour {
+		
+		public Camera cameraToPublish;
+		public string baseTopic = "/image_raw";
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/ImagePublisher.cs]: Please set default value to "/image_raw/compressed" and remove +"/compressed" in line 29 below.

File: Unity3D/Assets/RosSharp/Scripts/TransformManager.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
+﻿using UnityEngine;
+using System.Collections;
+
+namespace RosSharp
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/TransformManager.cs]: please change namespace to RosSharp.RosBridgeClient

File: Unity3D/Assets/RosSharp/Scripts/TransformManager.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+﻿using UnityEngine;
+using System.Collections;
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/TransformManager.cs]: Please rename the variables as follows:
linear_velocity -> linearVelocity
angular_velocity -> angularVelocity

File: Unity3D/Assets/RosSharp/Scripts/TransformManager.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace RosSharp
+{
+	public class TransformManager : MonoBehaviour
+	{
+		public bool useForceTorque = false;
+
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/TransformManager.cs]: Please remove useForceTorque and doUpdate.
Please also remove the comments corresponding to doUpdate below.

File: Unity3D/Assets/RosSharp/Scripts/TransformManager.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+﻿using UnityEngine;
+using System.Collections;
+
+namespace RosSharp
+{
+	public class TransformManager : MonoBehaviour
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/TransformManager.cs]: Please rename TransformManager to VelocityTransformManager

File: Unity3D/Assets/RosSharp/Scripts/VelocitySubscriber.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+﻿using UnityEngine;
+using System.Collections;
+
+namespace RosSharp.RosBridgeClient
+{
+
+	[RequireComponent(typeof(RosConnector))]
+	public class VelocitySubscriber : MonoBehaviour
+	{
+		public GameObject Model;
```
Comment (MartinBischoff) [inline Unity3D/Assets/RosSharp/Scripts/VelocitySubscriber.cs]: Please rename the variables as follows:
* Model -> UrdfModel
* twist_msg -> geometryTwist
* updateCmdVel -> updateVelocity

File: Unity3D/ROS/joy_to_vel.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import rospy
```
Comment (MartinBischoff) [inline Unity3D/ROS/joy_to_vel.py]: Please create a folder '/ROS/joy_publisher' in the root this repository and move both files there.
We will move the file_server and the Wiki launch files in this directory later.

File: Unity3D/ROS/turtlebot.launch [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<launch>
```
Comment (MartinBischoff) [inline Unity3D/ROS/turtlebot.launch]: Please rename this file to joy_to_vel.launch or a more generic name.

**Comment by jeremyfix** [comment]
Ok, I should have fixed your points. I definitely understand the critics. Let me know if other points should be fixed.

**Comment by jeremyfix** [comment]
One point about performance; By experimenting on some scenes, I noticed that rendering says turns at 30 FPS in the Editor but the image topic is published at a frame rate of 10 FPS. I'm not sure for now where there is this gap; 

Also, as speaking about performance, I'm not completely sure of the way I render the image before publishing it ... Blame of me, I took the code on a Unity forum but in the meanwhile, I noticed there are some other callbacks in MonoBehaviour. Maybe it might make sense to use another callback in CameraImageManager, avoiding to re-render the image ?!

**Comment by MartinBischoff** [comment]
Including CameraImagePublisher and VelocitySubscriber from @jeremyfix

**Pull request merged**